### ISSUE TITLE
[DAGCombine] Failure to drop SETCC in (x&1)!=0

### ISSUE THREAD

**Issue by LebedevRI** [llvm/llvm-project#50654]
|  |  |
| --- | --- |
| Bugzilla Link | [51312](https://llvm.org/bz51312) |
| Version | trunk |
| OS | Linux |
| CC | @topperc,@RKSimon,@phoebewang,@rotateright |

## Extended Description 
https://godbolt.org/z/aMq33KcGz

```
Optimized lowered selection DAG: %bb.0 'src:'
SelectionDAG has 13 nodes:
  t0: ch = EntryToken
            t2: v32i8,ch = CopyFromReg t0, Register:v32i8 %0
            t32: v32i8 = BUILD_VECTOR Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>
          t33: v32i8 = shl t2, t32
        t25: i32 = X86ISD::MOVMSK t33
      t26: i32 = parity t25
    t28: i8 = truncate t26
  t21: ch,glue = CopyToReg t0, Register:i8 $al, t28
  t22: ch = X86ISD::RET_FLAG t21, TargetConstant:i32<0>, Register:i8 $al, t21:1


Optimized lowered selection DAG: %bb.0 'tgt:'
SelectionDAG has 16 nodes:
  t0: ch = EntryToken
              t2: v32i8,ch = CopyFromReg t0, Register:v32i8 %0
              t24: v32i8 = BUILD_VECTOR Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>, Constant:i8<7>
            t25: v32i8 = shl t2, t24
          t20: i32 = X86ISD::MOVMSK t25
        t18: i32 = parity t20
      t10: i1 = setcc t18, Constant:i32<0>, setne:ch
    t11: i8 = any_extend t10
  t14: ch,glue = CopyToReg t0, Register:i8 $al, t11
  t15: ch = X86ISD::RET_FLAG t14, TargetConstant:i32<0>, Register:i8 $al, t14:1
```

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D117983

**Comment by RKSimon** [comment]
Fixed accc07e65465094dc5e12e78bee45b4d459c4ccd

**Comment by LebedevRI** [comment]
Thank you!
